test-engineering/contract-tests/partner/app/record_keeper.py

```diff
@@ -0,0 +1,41 @@
+from multiprocessing.managers import SyncManager
+from typing import Dict, List
+
+from fastapi import Request
+
+from models import Header, Record, Records
+
+
+class RecordKeeper:
+    """Responsible for Contile request history management"""
+
+    def __init__(self, multi_process_manager: SyncManager) -> None:
+        """Create an instance of RecordKeeper."""
+
+        self._records: List[Record] = multi_process_manager.list()
+
+    def add(self, request: Request) -> None:
+        """Create record from Fast API Request and add record to the record keeper."""
+
+        headers: List[Header] = [
+            Header(name=header[0], value=header[1])
+            for header in request.headers.items()
+        ]
+        query_parameters: Dict = dict(request.query_params)
```

Comment:
Based on what I'm seeing in the FastAPI and Starlette source code `request.query_params` returns an immutable multi-dict. Is it safe to cast that into a regular dict? Do you mind posting an example value here?

Reply:
Example Value: 
<img width="1353" alt="image" src="https://user-images.githubusercontent.com/3422688/178598900-0c5e82ec-289c-4f8e-bbe6-8c52c81687f6.png">

After some experimentation with the QueryParams, I can't confidently say that it is safe to explicitly cast to dict. Furthermore, using some of the onboard methods to try and create a to_dict() method was looking like overkill. 

[427b208](https://github.com/mozilla-services/contile/pull/415/commits/427b2085dbbcadd960e8f9f678df176585829577) - QueryParams is derived from request.url.query, a string. This raw form of the query, in my opinion, seems like the most accurate and appropriate form to assert against for test, so I opted to replace the `query_parameters:Dict` with `query:str`.

Reply:
[14528d7](https://github.com/mozilla-services/contile/pull/415/commits/14528d72ccbed458519a388d5fa4091727666575) - As discussed, the concern with casting to dict, is that values with the same key will not all be represented. The last appearance of a key-value pair will be returned. It is more practical to document expected values in a key-value structure and the appearance of duplicate keys is not an expected use case. 

For now we will cast explicitly and a task has been added to flag duplicate key values at the partner level. [CONSVC-1887](https://mozilla-hub.atlassian.net/browse/CONSVC-1887)